enterprise/server/remote_execution/filecache/filecache.go

```diff
@@ -298,15 +288,27 @@ func (c *fileCache) addFileToGroup(groupID string, node *repb.FileNode, existing
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	c.l.Remove(k)
-
 	fp := c.filecachePath(k)
-	if err := disk.EnsureDirectoryExists(filepath.Dir(fp)); err != nil {
-		return err
-	}
-	if err := cloneOrLink(groupID, existingFilePath, fp); err != nil {
-		return err
+
+	// During the initial directory scan (and only during the initial directory
+	// scan), the source and dest paths will be the same. In this case, make
+	// sure we don't evict an LRU entry that was added by some action before we
+	// got to this file in the scan. We'd wind up losing the cache entry in this
+	// case, because the eviction will unlink the destination file path (via the
+	// eviction callback), which is also the source file in this case.
+
+	if fp != existingFilePath {
+		c.l.Remove(k)
+		if err := disk.EnsureDirectoryExists(filepath.Dir(fp)); err != nil {
+			return err
+		}
+		if err := cloneOrLink(groupID, existingFilePath, fp); err != nil {
+			return err
+		}
+	} else if c.l.Contains(k) {
+		return nil
 	}
```

Comment:
this logic is a little confusing to me -- would it be clearer if we just enumerated the special case and unindented the rest?
```
if fp == existingFilePath && c.l.Contains(k) {
  return nil // skip adding the same file twice which could cause an eviction
}
c.l.Remove(k)
if err := disk.EnsureDirectoryExists(filepath.Dir(fp)); err != nil {
	return err
}
if err := cloneOrLink(groupID, existingFilePath, fp); err != nil {
	return err
}
```

Reply:
Done, but guarded the `Remove()`, `EnsureDirectoryExists()`, and `cloneOrLink()` calls with `if fp != existingFilePath { }`